src/Microsoft.VisualStudio.Threading/AsyncLazy`1.cs

```diff
@@ -234,6 +257,110 @@ public T GetValue(CancellationToken cancellationToken)
+            }
+        }
+    }
+
+    /// <summary>
+    /// Disposes of the lazily-initialized value if disposable, and causes all subsequent attempts to obtain the value to fail.
+    /// </summary>
+    /// <returns>
+    /// A task that completes when the value has been disposed of, or immediately if the value has already been disposed of or has been scheduled for disposal by a prior call.
+    /// </returns>
+    /// <remarks>
+    /// <para>Calling this method will put this object into a disposed state where future calls to obtain the value will throw <see cref="ObjectDisposedException"/>.</para>
+    /// <para>If the value has already been produced and implements <see cref="IDisposable"/> or <see cref="IAsyncDisposable"/>, it will be disposed of.
+    /// If the value factory has already started but has not yet completed, its value will be disposed of when the value factory completes.</para>
+    /// <para>If prior calls to obtain the value are in flight when this method is called, those calls <em>may</em> complete and their callers may obtain the value, although <see cref="IDisposable.Dispose"/>
+    /// may have been or will soon be called on the value, leading those users to experience a <see cref="ObjectDisposedException"/>.</para>
+    /// <para>Note all conditions based on the value implementing <see cref="IDisposable"/> or <see cref="IAsyncDisposable"/> is based on the actual value, rather than the <typeparamref name="T"/> type argument.
+    /// This means that although <typeparamref name="T"/> may be <c>IFoo</c> (which does not implement <see cref="IDisposable"/>), the concrete type that implements <c>IFoo</c> may implement <see cref="IDisposable"/>
+    /// and thus be treated as a disposable object as described above.</para>
+    /// </remarks>
+    public Task DisposeValueAsync()
+    {
+        Task<T>? localValueTask = null;
+        IDisposable? localValue = default;
+        lock (this.syncObject)
+        {
+            if (this.value == DisposedSentinel)
+            {
+                return Task.CompletedTask;
+            }
+
+            switch (this.value?.Status)
+            {
+                case TaskStatus.RanToCompletion:
+                    // We'll dispose of the value inline, outside the lock.
+                    localValue = this.value.Result as IDisposable;
+                    break;
+                case TaskStatus.Faulted:
+                case TaskStatus.Canceled:
+                    // Nothing left to do.
+                    break;
+                default:
+                    // We'll schedule the value for disposal outside the lock so it can be synchronous with the value factory,
+                    // but will not execute within our lock.
+                    localValueTask = this.value;
+                    break;
+            }
+
+            // Shut out all future callers from obtaining the value.
+            this.value = DisposedSentinel;
+
```

Comment:
I think we need add a memory barrier between line 334 and line 338.  I don't think the lock ensures execution inside always in the order, but only between statements before/after the lock. Here we actually depend on this.value to be set before valueFactory to be cleared out in another property, which do not use lock and depends on the order of setting those two values.